narwhal/primary/src/state_handler.rs

```diff
@@ -185,16 +250,184 @@ impl RandomnessState {
+                return;
+            }
+        };
+        let shares = match &dkg_output.shares {
+            Some(shares) => shares,
+            None => return, // can't participate in randomness generation without shares
+        };
+        debug!(
+            "random beacon: sending partial signatures for round {}",
+            self.randomness_round
+        );
+        let sigs = ThresholdBls12381MinSig::partial_sign_batch(
+            shares,
+            self.randomness_round.to_be_bytes().as_slice(),
+        );
+        let next_leader_narwhal_round = self.narhwal_round + (self.narhwal_round % 2);
+        self.last_narwhal_round_sent = next_leader_narwhal_round;
+
+        let leader = self.leader_schedule.leader(next_leader_narwhal_round);
+        if self.authority_id == leader.id() {
+            // We're the next leader, no need to send an RPC.
+            self.receive_partial_signatures(self.authority_id, self.randomness_round, sigs)
+                .await;
+            return;
+        }
+
+        let peer_id = anemo::PeerId(leader.network_key().0.to_bytes());
+        let peer = self.network.waiting_peer(peer_id);
+        let mut client = PrimaryToPrimaryClient::new(peer);
+        const SEND_PARTIAL_SIGNATURES_TIMEOUT: Duration = Duration::from_secs(10);
+        let request = anemo::Request::new(SendRandomnessPartialSignaturesRequest {
+            round: self.randomness_round,
+            sigs,
+        })
+        .with_timeout(SEND_PARTIAL_SIGNATURES_TIMEOUT);
+
+        if let Some(task) = &self.partial_sig_sender {
+            // Cancel previous partial signature transmission if it's not yet complete.
+            task.abort();
+        }
+        self.partial_sig_sender = Some(spawn_logged_monitored_task!(
+            async move {
+                let resp = client.send_randomness_partial_signatures(request).await;
+                if let Err(e) = resp {
+                    info!(
+                        "random beacon: error sending partial signatures to leader {leader:?}: {e:?}"
+                    );
+                }
+            },
+            "RandomnessSendPartialSignatures"
+        ));
+    }
+
+    async fn receive_partial_signatures(
+        &mut self,
+        authority_id: AuthorityIdentifier,
+        round: u64,
+        sigs: Vec<RandomnessPartialSignature>,
+    ) {
+        let dkg_output = match &self.dkg_output {
+            Some(dkg_output) => dkg_output,
+            None => {
+                error!("random beacon: called receive_partial_signatures before DKG completed");
+                return;
             }
+        };
+        if round < self.randomness_round {
+            debug!(
+                "random beacon: ignoring partial signatures for old round {round} (we are at {})",
+                self.randomness_round
+            );
+            return;
+        }
+        const MAX_ROUND_DELTA: u64 = 2;
+        if round > self.randomness_round + MAX_ROUND_DELTA {
```

Comment:
Can you pls add a comment about when we might see here round = randomness_round + 2? i'm guessing it might happen if the current validator is behind the rest of the network, and didn't see the last randomness round. If so, we should we limit it only by 2 rounds?